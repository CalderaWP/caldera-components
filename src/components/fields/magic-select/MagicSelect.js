import React from 'react';
import {
	onValueChangePropType,
	optionsShapeProp,
	valuePropType
} from '../propTypes';
import Autocomplete from 'react-autocomplete';
import PropTypes from 'prop-types';
import {MagicItem} from './MagicItem';

/**
 * Encapsulates a complete Magic Select field
 */
export class MagicSelect extends React.PureComponent {

	/**
	 * Create a MagicSelect component
	 *
	 * @param {Object} props
	 */
	constructor(props) {
		super(props);
		this.onChange = this.onChange.bind(this);
		this.onChange = this.onChange.bind(this);
		this.onSelect = this.onSelect.bind(this);
		this.onInputFocus = this.onInputFocus.bind(this);
		this.renderItem = this.renderItem.bind(this);
		this.onInputBlur = this.onInputBlur.bind(this);
	}

	/**
	 * Handle direct change events
	 *
	 * Probably not needed
	 * @param {Event} event
	 */
	onChange(event) {
		this.props.onValueChange(event.target.value);
	}

	/**
	 * Handle when the field gets focus
	 */
	onInputFocus() {
		if( 'function' === typeof this.props.onFocus ){
			this.props.onFocus();
		}
	}

	/**
	 * Handle when the field is blurred
	 */
	onInputBlur() {
		if( 'function' === typeof this.props.onBlur ){
			this.props.onBlur();
		}
	}


	/**
	 * Handle when the option is chosen
	 * @param {String|number} value
	 */
	onSelect(value) {
		this.props.onValueChange(value);
	}


	/**
	 * Render option
	 *
	 * @param {object} item
	 * @param {bool} isHighlighted
	 * @return {*}
	 */
	renderItem(item, isHighlighted){
		return <MagicItem item={item} isHighlighted={isHighlighted} innerKey={item.innerKey} key={item.innerKey} />
	}


	/**
	 * Render MagicSelect component
	 * @return {*}
	 */
	render() {
		return (

			<Autocomplete
				getItemValue={(item) => item.value}
				items={this.props.options}
				inputProps={{
					id: this.props.id,
					value: this.props.value,
					className: `${this.props.id}-magic-input`,
					onFocus: this.onInputFocus,
					onBlur:this.onInputBlur,
					onClick:this.onChange
				}}
				renderItem={this.renderItem}
				value={this.props.value}
				onChange={this.onChange}
				open={this.props.isOpen}
				selectOnBlur={true}
				onSelect={this.onSelect}
			/>

		);
	}
}


/**
 * Prop definitions for MagicSelect component
 */
MagicSelect.propTypes = {
	id: PropTypes.string.isRequired,
	options: optionsShapeProp,
	isRequired: PropTypes.bool,
	help: PropTypes.string,
	value: valuePropType,
	onValueChange: onValueChangePropType,
	disabled: PropTypes.bool,
	isOpen: PropTypes.bool,
	onBlur: PropTypes.func,
	onFocus: PropTypes.func
};

/**
 * Default property values for MagicSelect component
 *
 * @type {{}}
 */
MagicSelect.defaultProps = {
	defaultList: 'fields',
	isOpen: false,
	options:[]
};