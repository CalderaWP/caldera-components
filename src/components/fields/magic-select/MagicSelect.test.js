import renderer from 'react-test-renderer';
import React from 'react';
import {mount} from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {MagicSelect} from './MagicSelect';

Enzyme.configure({adapter: new Adapter()});

const genericChangeHandler = () => {
};
describe('MagicSelect component', () => {

	it('Matches snapshot', () => {
		const component = renderer.create(
			<MagicSelect
				id={'magic-1'}
				fieldClassName={'magic'}
				onValueChange={genericChangeHandler}
				options={[
					{
						label: 'HTML',
						value: 'html',
						innerKey:'html'
					},
					{
						label: 'Plain Text',
						value: 'plain',
						innerKey:'plain'
					}
				]}
			/>
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('Has inner input', () => {
		const component = mount(
			<MagicSelect
				id={'magic-2a'}
				fieldClassName={'magic'}
				onValueChange={genericChangeHandler}
				options={[
					{
						label: 'HTML',
						value: 'html',
						innerKey:'html'
					},
					{
						label: 'Plain Text',
						value: 'plain',
						innerKey:'plain'
					}
				]}
			/>
		);
		expect(component.find('input')).toHaveLength(1);
	});

	it('puts the right id attribute on inner input', () => {
		const component = mount(
			<MagicSelect
				id={'magic-2'}
				fieldClassName={'magic'}
				onValueChange={genericChangeHandler}
				options={[
					{
						label: 'HTML',
						value: 'html',
						innerKey:'html'
					},
					{
						label: 'Plain Text',
						value: 'plain',
						innerKey:'plain'
					}
				]}
			/>
		);
		expect(component.find('input').prop('id')).toBe('magic-2');
	});

	describe('Options', () => {
		it('shows no options if closed', () => {
			const component = mount(
				<MagicSelect
					id={'magic-3'}
					fieldClassName={'magic'}
					onValueChange={genericChangeHandler}
					options={[
			{
				label: 'HTML',
				value: 'html',
				innerKey:'html'
			},
			{
				label: 'Plain Text',
				value: 'plain',
				innerKey:'plain'
			}
		]}
					isOpen={false}
				/>
			);
			expect(component.find('.magic-input-option')).toHaveLength(0);
		});

		it('shows  options if opened', () => {
			const component = mount(
				<MagicSelect
					id={'magic-3'}
					fieldClassName={'magic'}
					onValueChange={genericChangeHandler}
					options={[
			{
				label: 'HTML',
				value: 'html',
				innerKey:'html'
			},
			{
				label: 'Plain Text',
				value: 'plain',
				innerKey:'plain'
			}
		]}
					isOpen={true}
				/>
			);
			expect(component.find('.magic-input-option')).toHaveLength(2);
		});

		it('Updates is open state when focused', () => {
			let isOpen = false;
			const component = mount(
				<MagicSelect
					id={'magic-9-blur'}
					fieldClassName={'magic'}
					onInputBlur={() => {
						isOpen = true;
					}}
					onFocus={() => {
						isOpen = false;
					}}
					onValueChange={genericChangeHandler}
					options={[
						{
							label: 'a1',
							value: 1,
							innerKey: '1'
						},
						{
							label: 'a14',
							value: 14,
							innerKey: '2'

						}
					]}
				/>
			);
			component.find( 'input' ).simulate('focus');
			expect(component.prop('isOpen')).toEqual(false);
		});

		it('Updates is open state when blurred', () => {
			let isOpen = false;
			const component = mount(
				<MagicSelect
					id={'magic-9-blur'}
					fieldClassName={'magic'}
					onValueChange={genericChangeHandler}
					onInputBlur={() => {
						isOpen = true;
					}}
					onFocus={() => {
						isOpen = false;
					}}
					options={[
						{
							label: 'a1',
							value: 1,
							innerKey: '1'
						},
						{
							label: 'a14',
							value: 14,
							innerKey: '2'

						}
					]}
				/>
			);
			component.find( 'input' ).simulate('focus');
			component.find( 'input' ).simulate('blur');
			expect(component.prop('isOpen')).toEqual(false);
		});

		it('Uses options prop by default - right number of options', () => {
			const component = mount(
				<MagicSelect
					id={'magic-4'}
					fieldClassName={'magic'}
					onValueChange={genericChangeHandler}
					options={[
						{
							label: 'a1',
							value: 1,
							innerKey: '1'
						},
						{
							label: 'a14',
							value: 14,
							innerKey: '2'

						}
					]}
					isOpen={true}
				/>
			);
			expect(component.find('.magic-input-option')).toHaveLength(2);
		});

		it('Uses options prop by default - right number of options', () => {
			const component = mount(
				<MagicSelect
					id={'magic-4'}
					fieldClassName={'magic'}
					onValueChange={genericChangeHandler}
					options={[
			{
				label: 'HTML',
				value: 'html',
				innerKey:'html'
			},
			{
				label: 'Plain Text',
				value: 'plain',
				innerKey:'plain'
			}
		]}
					isOpen={true}
				/>
			);
			expect(component.find('.magic-input-option')).toHaveLength(2);
		});

	});


	describe('Value', () => {
		it('Uses props.value to set value', () => {
			const component = mount(
				<MagicSelect
					id={'magic-7'}
					fieldClassName={'magic'}
					onValueChange={genericChangeHandler}
					options={[
						{
							label: '1',
							value: 1,
							innerKey: '11'
						}
					]}
					value={1}
				/>
			);
			expect(component.find('input').prop('value')).toBe(1);
		});


		it('Passes updated value properly through the onSelect handler', () => {
			let updatedValue = null;
			const component = mount(
				<MagicSelect
					id={'magic-9'}
					fieldClassName={'magic'}
					onValueChange={(newValue) => {
						updatedValue = newValue;
					}}
					options={[
						{
							label: '1',
							value: 1,
							innerKey: '12'
						},
						{
							label: '14',
							value: 14,
							innerKey: '13'

						}
					]}
					value={1}
				/>
			);

			component.instance().onSelect(14);
			expect(updatedValue).toEqual(14);
		});

		it.skip('Receives the updated value ', () => {
			let updatedValue = null;
			const component = mount(
				<MagicSelect
					id={'magic-9'}
					fieldClassName={'magic'}
					onValueChange={(newValue) => {
						updatedValue = newValue;
					}}
					options={[
						{
							label: '1',
							value: 12,
							innerKey: '12'
						}
					]}
					value={updatedValue}
					isOpen={true}
				/>
			);

			component.find('input').simulate('change', {event:{target: {value: 12}}} );
			expect(updatedValue).toEqual(14);
		});
	});



});