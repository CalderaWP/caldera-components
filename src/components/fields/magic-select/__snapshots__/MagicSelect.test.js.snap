// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MagicSelect component Matches snapshot 1`] = `
<div
  style={
    Object {
      "display": "inline-block",
    }
  }
>
  <input
    aria-autocomplete="list"
    aria-expanded={false}
    autoComplete="off"
    className="magic-1-magic-input"
    id="magic-1"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    role="combobox"
    value=""
  />
</div>
`;
